docs/recipes/UniversalE2ETesting.md

```diff
@@ -0,0 +1,464 @@
+ * import playwright code into a detox environment, or vice versa.
+ */
+export const test: Test = (() => {
+  const testEnvsLoaded = [isDetoxTestEnv(), isPlaywrightTestEnv()].filter(Boolean).length
+
+  if (testEnvsLoaded !== 1) {
+    throw new Error(
+      `${testEnvsLoaded} test environments loaded. Only one is allowed. Check the isTestEnv functions to make sure they check for globals that are specific only to their test environment`,
+    )
+  }
+
+  if (isDetoxTestEnv()) {
+    return require("./detox/entry").test
+  }
+  if (isPlaywrightTestEnv()) {
+    return require("./playwright/entry").test
+  }
+
+  throw new Error(
+    "Unknown test environment. Check the isTestEnv functions to make sure they check for globals that are specific only to their test environment",
+  )
+})()
+```
+
+### Page Object Interfaces – `e2e/screens.ts`
+
+Define your “agnostic” page objects as TypeScript interfaces. You should eventually have a page object model interface for each screen in your app, but for now we just have a login screen and a welcome screen.
+
+```ts title="e2e/screens.ts"
+export interface ILoginScreen {
+  login(): Promise<void>
+}
+
+export interface IWelcomeScreen {
+  launch(): Promise<void>
+}
+
+/** A fixture of all the page object models we can use in our tests */
+export type Fixtures = {
+  loadApp: () => Promise<void>
+  loginScreen: ILoginScreen
+  welcomeScreen: IWelcomeScreen
+  // ... you can add more as needed
+}
+```
+
+### Detox Implementation – `e2e/detox/screens/LoginScreen.ts`
+
+Behold! An implementation! This is where we implement the actual logic of our page object model. Because we are in the detox directory and we are careful loading only one test environment at a time in our `e2e/entry.ts` file, we can import Detox specific code here.
+
+```ts title="e2e/detox/screens/LoginScreen.ts"
+import { expect, element, by } from "detox"
+import type { ILoginScreen } from "../../screens"
+
+export class DetoxLoginScreen implements ILoginScreen {
+  async login() {
+    await expect(element(by.text("Log In"))).toBeVisible()
+    await element(by.text("Tap to log in!")).tap()
+  }
+}
+```
+
+### Detox Implementation – `e2e/detox/screens/WelcomeScreen.ts`
+
+Behold! Another implementation!
+
+```ts title="e2e/detox/screens/WelcomeScreen.ts"
+import { expect, element, by } from "detox"
+import type { IWelcomeScreen } from "../../screens"
+
+export class DetoxWelcomeScreen implements IWelcomeScreen {
+  async launch() {
+    await expect(element(by.text("Your app, almost ready for launch!"))).toBeVisible()
+    await element(by.text("Let's go!")).tap()
+    await expect(element(by.text("Components to jump start your project!"))).toBeVisible()
+  }
+}
+```
+
+### Detox Setup – `e2e/detox/setup.ts`
+
+This file has a lot of logic, but this is where we load the app in a way that is compatible with both debug and release builds. You can customize this to your liking, but this is a good starting point.
```

Comment:
Is it important to know what the logic is doing here?

Reply:
The logic here is to do the right setup steps based on whether it's a local development build, a debug build, or an EAS build: I got it from https://github.com/expo/eas-tests-example/blob/main/e2e/tests/utils/openApp.js

It might be better to explain that more explicitly then

Reply:
[docs: Enhance Detox setup explanation in E2E testing recipe](https://github.com/infinitered/ignite-cookbook/pull/186/commits/6a2aec212e34df6346b405c79376b36dcf2bf534)